x-pack/plugins/security/server/authentication/index.ts

```diff
@@ -105,28 +105,25 @@ export async function setupAuthentication({
       // authentication (username and password) or arbitrary external page managed by 3rd party
       // Identity Provider for SSO authentication mechanisms. Authentication provider is the one who
       // decides what location user should be redirected to.
-      return t.redirected(authenticationResult.redirectURL!);
+      return response.redirected(undefined, {
+        headers: {
+          location: authenticationResult.redirectURL!,
+        },
+      });
     }
 
     let error;
     if (authenticationResult.failed()) {
       authLogger.info(`Authentication attempt failed: ${authenticationResult.error!.message}`);
       error = wrapError(authenticationResult.error);
-
-      const authResponseHeaders = authenticationResult.authResponseHeaders;
-      for (const [headerName, headerValue] of Object.entries(authResponseHeaders || {})) {
-        if (error.output.headers[headerName] !== undefined) {
-          authLogger.warn(`Server rewrites a error response header [${headerName}].`);
-        }
-        // Hapi typings don't support headers that are `string[]`.
-        error.output.headers[headerName] = headerValue as any;
-      }
     } else {
       authLogger.info('Could not handle authentication attempt');
       error = Boom.unauthorized();
     }
 
-    return t.rejected(error);
+    return response.unauthorized(error, {
+      headers: authenticationResult.authResponseHeaders,
```

Comment:
should we add headers potentially set in `authenticationResult.error`
as I can see ` authenticationResult.authResponseHeaders` already contains `WWW-Authenticate`

@elastic/kibana-security

Reply:
> should we add headers potentially set in authenticationResult.error

I don't think so. Authentication providers can leverage `authResponseHeaders` and should be minimally dependent on Boom internals, imo. If core provides its own error helpers/types someday we'll happily migrate to them (we just use `Boom.badRequest(string)` and `Boom.unauthorized()` right now).

Reply:
I didn't mean to handle case when error is Boom object with set headers. My question whether we Security plugin set some headers on Boom error and we should do here
```js
  return response.unauthorized(error, {
     ...(error.output.headers || {})
      headers: authenticationResult.authResponseHeaders,
  })
```

Reply:
```js
  return response.unauthorized(error, {
     ...(error.output.headers || {})
     headers: authenticationResult.authResponseHeaders,
  })
```

Hmm, yeah, that's what I replied to :slightly_smiling_face: We (security) shouldn't set any headers on error object since we have `authResponseHeaders` for that purpose.